Stackless/module/scheduling.c

```diff
@@ -710,38 +710,69 @@ new_lock(void)
  * whereas Stackless stores the exception state in the tasklet object.
  * When switching from one tasklet to another tasklet, we have to switch
  * the exc_info-pointer in the thread state.
+ *
+ * The same situation is for the current contextvars.Context. When switching
+ * from one tasklet to another tasklet, we have to switch the context-pointer
+ * in the thread state.
  */
 
+#if 1
+Py_LOCAL_INLINE(void) SLP_EXCHANGE_EXCINFO(PyThreadState *tstate, PyTaskletObject *task)
+{
+    PyThreadState *ts_ = (tstate);
+    PyTaskletObject *t_ = (task);
+    _PyErr_StackItem *exc_info;
+    PyObject *c;
+    assert(ts_);
+    assert(t_);
+    exc_info = ts_->exc_info;
+    assert(exc_info);
+    assert(t_->exc_info);
 #if 0
-#define SLP_EXCHANGE_EXCINFO(tstate, task) \
+    c = PyStackless_GetCurrent();
+    fprintf(stderr, "SLP_EXCHANGE_EXCINFO %3d current %14p,\tset task %p = %p,\ttstate %p = %p\n", __LINE__, c, t_, exc_info, ts_, t_->exc_info);
+    Py_XDECREF(c);
+#endif
+    ts_->exc_info = t_->exc_info;
+    t_->exc_info = exc_info;
+    c = ts_->context;
+    ts_->context = t_->context;
+    t_->context = c;
+    ts_->context_ver++;
+}
+#else
+#define SLP_EXCHANGE_EXCINFO(tstate_, task_) \
     do { \
-        PyThreadState *ts_ = (tstate); \
-        PyTaskletObject *t_ = (task); \
+        PyThreadState *ts_ = (tstate_); \
+        PyTaskletObject *t_ = (task_); \
         _PyErr_StackItem *exc_info; \
-        PyObject * c = PyStackless_GetCurrent(); \
+        PyObject *c; \
         assert(ts_); \
         assert(t_); \
         exc_info = ts_->exc_info; \
         assert(exc_info); \
         assert(t_->exc_info); \
-        fprintf(stderr, "SLP_EXCHANGE_EXCINFO %3d current %14p,\tset task %p = %p,\ttstate %p = %p\n", __LINE__, c, t_, exc_info, ts_, t_->exc_info); \
-        Py_XDECREF(c); \
         ts_->exc_info = t_->exc_info; \
         t_->exc_info = exc_info; \
+        c = ts_->context; \
+        ts_->context = t_->context; \
+        t_->context = c; \
+        ts_->context_ver++; \
```

Comment:
See previous comment about why the currently runnint tasklet has a "saved" context?  Also, what is `context_ver`?

Reply:
Abaut context_ver: 

A context is a Mapping from ContextVar objects to their values. A ContextVar caches its current value and the current context_ver. To invalidate the cache context_ver must be incremented on each change of ts->context. See https://github.com/stackless-dev/stackless/blob/4c7f02006d6fb0f2b5eb8fca449dd2563fa924ef/Python/context.c#L759